### DISCUSSION TITLE
How can ESP32 place microython startup on psram

### DISCUSSION THREAD

**Discussion by Zhengqinze05** [micropython/micropython#9964] [Core Development]
I use esp32S3. I have an idea about whether I can place the startup of microython in psram instead of iram. If so, what changes do I need to make? I want to save the space of iram as much as possible

**Comment by jimmo** [comment]
What do you mean by "the startup of micropython"

If something is in iram, that's typically because it has to be.

**Reply by Zhengqinze05** [reply]
After I used microPython, the iram of the system decreased by more than 100 K. However, I think other boards, such as the esp8266, do not need so much memory. Why does the esp32 have such a high iram utilization rate. The gc_init () function also uses the memory requested with psram, but iram does not know where it uses so much.

**Reply by jimmo** [reply]
Right. On the **ESP32**, code can either execute from flash or from IRAM, but _not_ PSRAM. It is much faster to execute code from IRAM, so MicroPython puts a few performance-critical functions into IRAM. This does make a fairly significant performance increase. See https://github.com/micropython/micropython/pull/7850 as a recent example of moving some more code into IRAM.

But this has nothing to do with the startup code for MicroPython. It's specific functions to do with the runtime (such as the VM).

However, on the **ESP32S3**, it is possible to execute from PSRAM. So yes, the ESP32 port could be configured to place these functions into PSRAM instead. See the link @Wind-stormger posted, as well as the more general guide https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/api-guides/external-ram.html#move-instructions-in-flash-to-psram

See https://github.com/micropython/micropython/blob/master/ports/esp32/mpconfigport.h#L182 which configures these functions to use `IRAM_ATTR`.

**Comment by Wind-stormger** [comment]
Check out this document if you have a very urgent need for this.

https://docs.espressif.com/projects/esp-idf/en/release-v4.4/esp32s3/api-guides/performance/ram-usage.html#optimizing-iram-usage
